### PR TITLE
Move command line params to TableGen version like in the current version of the llvm-objdump

### PR DESCRIPTION
**Pull request by sv99**
Move from llvm::cl to TableGen version command line check.

Add --sysroot params for working with Mac (my working system).

--include-file - add files from include in the depth - now checked anyway.

### CODE REVIEW

**Comment by sv99** [comment]
Please, view my PR. I have two next big PR with massive internal vars renaming for satisfy global clang-tidy style.

**Comment by bharadwajy** [comment]
Thanks! Good code enhancement proposal.

Following are the options related to specifying external function prototypes in the current repo. 

```
--include-files=<string>         - List of comma-seperated header files with function prototypes using standard C syntax.
  -I                             - Alias for --include-files=<single-header-file>
```

It appears that with this PR we've lost support for `--include-files`. I see `--include-file` but it appears different from `--include-files` which expects a list of comma-separated files.

Build target `check-mctoll` results in 4 test failures due to this.

Can you please address this?

**Comment by sv99** [comment]
I dont find flag --include-file in the current version llvm-mctoll.cpp

https://github.com/microsoft/llvm-mctoll/blob/9a289e20fe972219db98a512d40f5b92b48a94ca/llvm-mctoll.cpp#L182-L191

Which test filed?

**Comment by sv99** [comment]
For working with other targets I add --sysroot flag for toolchain path (different from default system path - /)

**Comment by sv99** [comment]
My working system is MAC - for compiling to linux target and arm target I made toolchain https://github.com/sv99/llvm-mctoll-toolchains

**Comment by sv99** [comment]
My working system is MAC - for compiling to linux target and arm target I made toolchain https://github.com/sv99/llvm-mctoll-toolchains

**Comment by sv99** [comment]
Change 4 tests

File: test/smoke_test/test-array-changed-in-callee.c [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // RUN: clang -o %t-opt %s
-// RUN: llvm-mctoll -d --include-files="/usr/include/stdio.h,/usr/include/stdlib.h" %t-opt
+// RUN: llvm-mctoll -d -I /usr/include/stdio.h -I /usr/include/stdlib.h %t-opt
```
Comment (bharadwajy) [inline test/smoke_test/test-array-changed-in-callee.c:2]: I had added `-I` as an alias for `--include-files=<single-header-file>` and works in the current `master` branch. I understand  your change  continues  to support it.

But, I'd like to keep the option `--include-files=<comma-separated-list-of-header-files>`. So  changing the test is not preferable.

Is there no way to support such options in `Opts.td`?

Comment (sv99) [reply test/smoke_test/test-array-changed-in-callee.c]: Flag added

**Comment by bharadwajy** [comment]
> For working with other targets I add --sysroot flag for toolchain path (different from default system path - /)
> 
> My working system is MAC - for compiling to linux target and arm target I made toolchain https://github.com/sv99/llvm-mctoll-toolchains

Can you add a (markdown) document in `doc` directory specifying the steps to be followed for cross compilation on MacOS? This will be of help to developers who might be interested in taking that route.

**Comment by sv99** [comment]
I have some changes at work to unify the test suite for working on linux and mac smoothly.
In my fork I have two massive changes for renaming internal vars for satisfy global clang-tidy and now I am working on the Arm - problems with calling function.

**Comment by bharadwajy** [comment]
> I have some changes at work to unify the test suite for working on linux and mac smoothly. In my fork I have two massive changes for renaming internal vars for satisfy global clang-tidy and now I am working on the Arm - problems with calling function.

That is great to learn. Since this change introduces a new option `--sysroot` is would be appropriate to add the necessary documentation to help users as part of this change. I'd suggest that each of the enhancements be self-contained to help during any potential retrospective for bug origins at a later time. 

This PR looks good to me, but would be great to add the documentation / usage expectation for the new option `--sysroot`.

**Comment by sv99** [comment]
doc file added

File: Opts.td [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  HelpText<"Operate on the specified sections only. "
+           "With --macho dump segment,section">,
+  Flags<[HelpHidden]>;
+def : Separate<["--"], "section">, Alias<section_EQ>;
+def : Separate<["-"], "j">, Alias<section_EQ>,
+  HelpText<"Alias for --section">,
+  Flags<[HelpHidden]>;
+
+def sysyroot_EQ : Joined<["--"], "sysroot=">,
+  HelpText<"Toolchain sysroot">;
```
Comment (bharadwajy) [inline Opts.td:94]:
```suggestion
  HelpText<"Logical root directory for headers specified using --include-files, if any. [default system toolchain dir]">;
```

File: Opts.td [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+def : Separate<["--"], "sysroot">, Alias<sysyroot_EQ>;
+
+def target_EQ : Separate<["-"], "target">,
+  HelpText<"Target triple to disassemble for, "
+            "see --version for available targets">,
+  Flags<[HelpHidden]>;
```
Comment (bharadwajy) [inline Opts.td:100]: Why is this needed? The target ISA is determined by the input binary. How is this user-specified option expected be consumed?

File: Opts.td (hunk 2 above)
Comment (bharadwajy) [reply Opts.td:94]: I also notice that `llvm-mctoll --help-hidden` prints the internal `--sysroot` option of LLVM AND this one as follows:

```
...
--sysroot=<value>       Toolchain sysroot
--sysroot <value>       Toolchain sysroot
...
```
Need to ensure there is no conflict and the option is not printed twice - which is confusing.

File: Opts.td [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+def start_address_EQ : Joined<["--"], "start-address=">,
+  MetaVarName<"address">,
+  HelpText<"Disassemble beginning at address">,
+  Flags<[HelpHidden]>;
+def : Separate<["--"], "start-address">, Alias<start_address_EQ>;
+def stop_address_EQ : Joined<["--"], "stop-address=">,
+  MetaVarName<"address">,
+  HelpText<"Stop disassembly at address">,
+  Flags<[HelpHidden]>;
+def : Separate<["--"], "stop-address">, Alias<stop_address_EQ>;
```
Comment (bharadwajy) [inline Opts.td:82]: I notice that `llvm-mctoll --help-hidden` prints the internal `--start-address` and `--stop-address` options of LLVM AND these as follows:

```
...
  --start-address=address Disassemble beginning at address
  --start-address <value> Disassemble beginning at address
  --stop-address=address  Stop disassembly at address
  --stop-address <value>  Stop disassembly at address
...
```
Need to ensure there is no conflict and these options are not printed twice - which is confusing.

File: Opts.td (hunk 2 above)
Comment (sv99) [reply Opts.td:97]: Doubled help messages hided.
The -target flag needed for specify toolchain (pair with --sysroot flag) for parsing header.
For arm binary calculated generalized triple, which no satisfy with real toolchain.

**Comment by sv99** [comment]
Strange conflict message! Do I need to do something?

**Comment by bharadwajy** [comment]
> Strange conflict message! Do I need to do something?

I pushed a bug fix exposed when I built the repo on Ubuntu 22.04. Please rebase the PR.

**Comment by sv99** [comment]
PR rebased

File: doc/sysroot_flag.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+llvm-mctoll --sysroot ~/toolchain/arm-linux-gnueabihf \
+  -target arm-linux-gnueabihf -I header-inc.h \
+  -debug -d hello-arm
+# ELF 64-bit x86_64 Linux
+lvm-mctoll --sysroot ~/toolchain/x86_64-linux-gnu \
+  -target x86_64-linux-gnu -I header-inc.h \
+  -debug -d hello-arm 
+ ```
+
+### run arm binary using docker
```
Comment (bharadwajy) [inline doc/sysroot_flag.md:54]:
```suggestion
### Running Linux ARM binary using docker on a non-Linux development host
```

**Review by bharadwajy — COMMENTED** [review]
Thanks for the changes.

Please look through the PR feedback and resolve them. Most of the suggestions are content edits.

**Comment by sv99** [comment]
doc/sysroot_flag.md corrected
Sorry for my english.

File: Raiser/IncludedFileInfo.cpp [hunk 6]
```diff
@@ -151,16 +151,16 @@ class FuncDeclFinder : public clang::ASTConsumer {
...
-    clang::SourceManager &SourceManager(Context.getSourceManager());
+    // clang::SourceManager &SourceManager(Context.getSourceManager());
     for (auto &Decl : Decls) {
       if (Decl->isFunctionOrFunctionTemplate()) {
-        const auto &FileID = SourceManager.getFileID(Decl->getLocation());
-        if (FileID != SourceManager.getMainFileID())
-          continue;
+        // const auto &FileID = SourceManager.getFileID(Decl->getLocation());
+        // if (FileID != SourceManager.getMainFileID())
+        //   continue;
```
Comment (bharadwajy) [inline Raiser/IncludedFileInfo.cpp:159]: Why was this commented out?

Comment (sv99) [reply Raiser/IncludedFileInfo.cpp:159]: With this lines scan function prototipes **only** in this file, internal #include visitor ignored.
But if defined sysroot and target and exists toolchain with headers tooling loads all internal includs.

Without this line may define header file like this

```bash
// header-inc.h

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
```

Comment (bharadwajy) [reply Raiser/IncludedFileInfo.cpp:159]: The code commented out gets the `FileID` of the candidate `Decl` - which `isFunctionOrFunctiontemplate`. The `if` condition checks to ensure `Decl` is visited (and considered for addition to the list of external/known function prototypes) only if it belongs to the main file (such as `/usr/inculde/stdio.h`).

I don't see how this code is involved in skipping over internal `#include`s. Can you please add some code here using `LLVM_DEBUG` (such as on [line 81](https://github.com/sv99/llvm-mctoll/blob/d45c4fae5147f12284191f8e18efabf1515c6f32/Raiser/IncludedFileInfo.cpp#L81) of this file) to demonstrate your observation. Thx!

Comment (sv99) [reply Raiser/IncludedFileInfo.cpp:159]: Problem with `if (FileID != SourceManager.getMainFileID())`
if using my  header-inc.h file (instead of `/usr/inculde/stdio.h`) list function is empty!!

With commented lines - include all files from  <stdio.h>,  <stdlib.h> and <string.h>

Comment (bharadwajy) [reply Raiser/IncludedFileInfo.cpp:159]: OK. I believe [this patch](https://github.com/microsoft/llvm-mctoll/files/9539276/incf.txt) addresses the issue appropriately. I've checked its correctness by 

a) building the target `check-mctoll` and 
b) specifying a `header-inc.h` that includes other include files, as you indicated.

Please apply it and let me know if you have any questions. Thx!

**Comment by sv99** [comment]
Patch applied

**Review by bharadwajy — APPROVED** [review]
